### PR TITLE
deferred (or async) sequence number assignment

### PR DESCRIPTION
**Pull request by ymatsuda**
@eishay @stkem @raymondng 

NOTE:
Seq# is set to -1 on insert/update for now due to column type issue. I will change this to (clock.now.getMillis() - Long.MaxValue) once the column type is fixed.

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/NormalizedURIRepo.scala [hunk 1]
```diff
@@ -120,7 +120,9 @@ extends DbRepo[NormalizedURI] with NormalizedURIRepo with ExternalIdColumnDbFunc
 
   override def save(uri: NormalizedURI)(implicit session: RWSession): NormalizedURI = {
     log.info(s"about to persist $uri")
-    val num = sequence.incrementAndGet()
+
+    // setting a negative sequence number for deferred assignment
+    val num = SequenceNumber[NormalizedURI](-1L) // (clock.now.getMillis() - Long.MaxValue)
```
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/model/NormalizedURIRepo.scala]: So we lose fairness in the near-term?

File: kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/MySQL.scala [hunk 2]
```diff
@@ -33,6 +33,28 @@ class MySQL(val masterDb: SlickDatabase, val slaveDb: Option[SlickDatabase])
       rs.next()
       SequenceNumber[T](rs.getLong(1))
     }
+
+    def reserve(n: Int)(implicit sess: RWSession): SequenceNumberRange[T] = {
```
Comment (raymondng) [inline kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/MySQL.scala:37]: Nice :+1:

**Comment by raymondng** [comment]
Overall lgtm. Thanks!

File: kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/MySQL.scala [hunk 3]
```diff
@@ -33,6 +33,28 @@ class MySQL(val masterDb: SlickDatabase, val slaveDb: Option[SlickDatabase])
...
+      if (n > 0) {
+        val ts = System.currentTimeMillis()
+        val stmt = sess.getPreparedStatement(s"UPDATE $name SET id=LAST_INSERT_ID(id+?);")
```
Comment (stkem) [inline kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/MySQL.scala:40]: Do we still need `LAST_INSERT_ID`? It's leader only and single threaded, so no-one else it touching this table, i.e. `UPDATE $name SET id=id+?;` would do the same. Similarly a few lines below you wouldn't need it either. Or am I missing something?

**Comment by stkem** [comment]
:+1:

File: kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/MySQL.scala (hunk 3 above)
Comment (ymatsuda) [inline kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/MySQL.scala:40]: I don't want to use Zookeeper for crucial mutex control. There is a chance that multiple servers think they are leaders when notifications are delayed. The most reliable mutex is the locking on the row in the sequence number table.

Comment (stkem) [inline kifi-backend/modules/sqldb/app/com/keepit/common/db/slick/MySQL.scala:40]: ah, makes sense

**Pull request merged**